### PR TITLE
Describe included blocks.

### PR DESCRIPTION
**Pull request by gkellogg**
For #19.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/pull/208.html" title="Last updated on Aug 2, 2019, 5:28 PM UTC (809409a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/208/ab5050d...809409a.html" title="Last updated on Aug 2, 2019, 5:28 PM UTC (809409a)">Diff</a>

### CODE REVIEW

**Comment by azaroth42** [comment]
LGTM! :)

**Comment by azaroth42** [comment]
(Sorry, misclick)

File: common/terms.html [hunk 1]
```diff
@@ -225,6 +225,9 @@
     A <a>named graph</a> created from the value of a <a>map entry</a>
     having an <a>expanded term definition</a>
     where <code>@container</code> is set to  <code>@graph</code>.</dd>
+  <dt class="changed"><dfn data-cite="JSON-LD11#dfn-included-block">included block</dfn></dt><dd class="changed">
+    An <a>included block</a> is an <a>entry</a> in a <a>node object</a> where the key is `@included` or an alias
```
Comment (dlongley) [inline common/terms.html]:
```suggestion
    An <a>included block</a> is an <a>entry</a> in a <a>node object</a> where the key is either `@included` or an alias of `@included`
```

File: index.html [hunk 2]
```diff
@@ -608,6 +608,10 @@ <h2>Syntax Tokens and Keywords</h2>
         Used in a <a>context definition</a> to load an external context
         within which the containing <a>context definition</a> is merged.
         This can be useful to add JSON-LD 1.1 features to JSON-LD 1.0 contexts.</dd>
+      <dt class="changed">`@included`</dt><dd class="changed">
+        Used in in a top-level <a>node object</a> to define an <a>included block</a>,
```
Comment (dlongley) [inline index.html]:
```suggestion
        Used in a top-level <a>node object</a> to define an <a>included block</a>,
```

File: index.html [hunk 3]
```diff
@@ -8272,6 +8276,150 @@ <h3>Using the Document Base for the Default Vocabulary</h3>
 </section>
 </section>
 
+<section class="informative changed"><h3>Included Nodes</h3>
+  <p>Sometimes it is useful to list node objects as part of another node object.
+    For instance, to represent a set of resources which are used by some other
+    resource.
+    In JSON-LD. these resources could all be contained as members of an array,
```
Comment (dlongley) [inline index.html]:
```suggestion
    In JSON-LD, these resources could all be contained as members of an array,
```

File: index.html [hunk 4]
```diff
@@ -11918,6 +12073,16 @@ <h2>Type Maps</h2>
       added as a <code>@type</code> of the <a>node object</a> value when expanding.</p>
   </section>
 
+  <section class="changed">
+    <h2>Included Blocks</h2>
+
+    <p>An <a>included block</a> is used to provide a set of <a>node objects</a>.
+      An <a>included block</a> MAY appear as the value of a member of a node object with the key of `@included`, or an alias.
```
Comment (dlongley) [inline index.html]:
```suggestion
      An <a>included block</a> MAY appear as the value of a member of a node object with either the key of `@included` or an alias of `@included`.
```

File: index.html [hunk 5]
```diff
@@ -12246,6 +12411,12 @@ <h2>Keywords</h2>
       This keyword is described further in <a class="sectionRef" href="#context-definitions"></a>,
       and <a class="sectionRef" href="#default-vocabulary"></a>.
     </dd>
+    <dt><code>@included</code></dt><dd>
+      The `@included` keyword MAY be aliased,
```
Comment (dlongley) [inline index.html]:
```suggestion
      The `@included` keyword MAY be aliased and
```

**Comment by iherman** [comment]
Can we simplify the example to contain only what is really necessary to understand the feature? From the point of view of this feature aliasing `@type`, `@id`, or indeed `@included`, but also the usage of `@base` seems to be totally irrelevant. The same for the usage of type for the included objects. 

This would help the user concentrating on what is really essential in this case...

(Otherwise it looks o.k. to me...)

**Comment by gkellogg** [comment]
The example came from Rob’s original use case, but I can simplify it easily enough.

**Comment by iherman** [comment]
Actually... the comment of @dlongley in https://github.com/w3c/json-ld-syntax/issues/128#issuecomment-516936845 made me realize that, at least for me, the most important aspect of this feature is not what is in the proposed text. Indeed, what `@included` can replace is the misleading top level `@graph` when what one wants to expess is a “bush”, ie, several top level objects with their (recursive) properties. This is used all over the place in Turtle data, and for which the only solution in JSON-LD 1.0 is to use a top-level `@graph` property whoise value is... _not_ a graph in the dataset sense (certainly not the same way as when a TriG data is encoded in JSON-LD).

If my understanding is indeed correct, I would think that such an example should be “the” core example for this feature, and the cross references in @azaroth42’s example is just a welcome consequence... 

Or, at least, this fact should be emphasized in the text imho.

**Comment by gkellogg** [comment]
@iherman Please suggest some text and an example to help flesh this out in the spec. I'll be happy to integrate it into the text.

**Comment by dlongley** [comment]
@iherman,

> Indeed, what @included can replace is the misleading top level @graph when what one wants to expess is a “bush”, ie, several top level objects with their (recursive) properties.

Yes, this is also my understanding -- I left a comment to similar affect over on the API PR:

https://github.com/w3c/json-ld-api/pull/128#pullrequestreview-269051034

**Comment by iherman** [comment]
@gkellogg 

> @iherman Please suggest some text and an example to help flesh this out in the spec. I'll be happy to integrate it into the text.

Hopefully, the PR over this PR works: https://github.com/w3c/json-ld-syntax/pull/209

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-08-02-json-ld#section4-1).

- `RESOLVED:  focus <code>@included</code> text and example on original inclusion use case; mention value of it as an <code>@graph</code> replacement for bushes–and reference primer for further reading`
- `RESOLVED:  close issue #19 with merger of <code>@included</code> related PRs`
<details><summary><i class="differentiate">View the transcript</i></summary>
4.1. Indexing with <code>@included</code><br/>
Ivan Herman: -> <a href="https://github.com/w3c/json-ld-syntax/pull/208">https://github.com/w3c/json-ld-syntax/pull/208</a> issue PR<br/>
Ivan Herman: -> <a href="https://github.com/w3c/json-ld-syntax/issues/19">https://github.com/w3c/json-ld-syntax/issues/19</a> issue itself<br/>
Benjamin Young: We made good progress on this last week.<br/>
Gregg Kellogg: It is <code>@included</code> now.16:13:56 &lt;bigbluehat> <a href="https://jsonapi.org/">https://jsonapi.org/</a><br/>
… <code>@included</code> comes from the JSON.API spec, and we are adopting this.<br/>
… Right now it’s just a container for collecting node objects that don’t have a direct rel with the node in which they are contained.<br/>
… There’s been some exchange on the issue highlighting a bush-like use for included.<br/>
… In JSON-LD 1.0 the top-level graph is used to collect nodes is a corner case. Everywhere else where graphs are used are seen as named graph.<br/>
… Included doesn’t carry that baggage.<br/>
… So <code>@included</code> can be used in favor of <code>@graph</code> in these places.<br/>
… In 1.0, you can’t have a graph name being a property of another node. With <code>@included</code> you can.<br/>
… We can’t use <code>@graph</code> to define a default graph.<br/>
… Except when it is the only property in a top-level object.<br/>
Benjamin Young: How does the actual inclusion take place?<br/>
Gregg Kellogg: The shape is similar to JSON.API. The value is seen as an array of node objects. If you have a node that is a value of a prop….<br/>
… Needed in jsonapi for node references that are not included in the main document as references, but should be included aside.<br/>
… Included blocks can be nested, and will be flattened out when done.<br/>
… You won’t compact back to included blocks after flattening.<br/>
… You can use included in a frame and have it match on diff subjects.<br/>
… The name <code>@included</code> is out of sync with other keywords.<br/>
… Dave suggested <code>@include</code><br/>
… Just like jsonapi<br/>
Ivan Herman: It’s becoming bikeshedding<br/>
… I would go further than what you did. Ex 1.1.1 and 1.1.2 (bushes) should be removed.<br/>
… We should convince people to not use those forms anymore.<br/>
Gregg Kellogg: <a href="https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/pull/208.html#example-111-using-graph-to-explicitly-express-the-default-graph">https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/pull/208.html#example-111-using-graph-to-explicitly-express-the-default-graph</a><br/>
Gregg Kellogg: By removing these we won’t lose anything. We would have to remove everything after the note.<br/>
Gregg Kellogg: <a href="https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/pull/208.html#example-103-simple-data-with-several-top-level-nodes">https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/pull/208.html#example-103-simple-data-with-several-top-level-nodes</a><br/>
Gregg Kellogg: Also, we may want to reverse example 103 and 104. To clarify writing bushes.<br/>
Ivan Herman: Referring to <code>@graph</code> is misleading, as it has not been explained yet there.<br/>
Gregg Kellogg: We may want to change an example in the <code>@graph</code> section then.<br/>
… We can also just leave that out and use it in the best practices document.<br/>
Ivan Herman: Yes<br/>
Benjamin Young: Flattened representation will still contain <code>@graph,</code> so readers will have to know what it does.<br/>
… This plumbing shift is significant.<br/>
… This is going to cause issues for the json people that are operating on the flattened form.<br/>
Gregg Kellogg: If you flatten with a context, it would introduce a graph to contain it. This would change the shape dramatically.<br/>
… Same with framing. In 1.1 we don’t use an <code>@graph</code> at the top level if not needed.<br/>
… We could change the algo to use included instead.<br/>
… But we may not want to do that.<br/>
… So do we want to replace the main usage of <code>@graph</code> to <code>@included?</code><br/>
… Included allows embedded nodes to go to one place. Like in jsonapi, they don’t want to include referenced things inline, but only a reference to an included block.<br/>
Benjamin Young: Useful for reducing payload size. And only including referenced things once.<br/>
… These are just IRI references?<br/>
Gregg Kellogg: There is no magic going on.<br/>
Ivan Herman: In JSON-LD it used to be hard to do these indexed references.<br/>
… The bush features can now be expressed in two different ways.<br/>
… These things happen.<br/>
… It’s a matter of taste which one you prefer.<br/>
… I personally always hated graph for representing bushes.<br/>
… I like this new included representation for bushes.<br/>
… I don’t want to hide the fact that bushes can be described with included instead of the graph ‘hack’.<br/>
Gregg Kellogg: We can say that included can also be used without other props in node object for describing node objects without semantic relationship.<br/>
Ivan Herman: I’m fine with that.<br/>
Benjamin Young: <a href="https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/pull/208.html#included-blocks-to-be-flattened">https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/pull/208.html#included-blocks-to-be-flattened</a><br/>
Benjamin Young: If <code>@included</code> were <code>@graph,</code> this would make a named graph?<br/>
Gregg Kellogg: I think this would make one or two named graphs.<br/>
Benjamin Young: With included it won’t make named graphs.<br/>
Gregg Kellogg: Yes, just objects.<br/>
Benjamin Young: I see the value, but not keen on the new keyword.<br/>
… I think we need to explain these next to each other, with their nuances.<br/>
… The initial reason for this feature was not meant to displace graph.<br/>
… It was meant to bring in other referenced objects in the document.<br/>
Gregg Kellogg: What jsonld always had was the ability to reference node<br/>
… by defining <code>@id</code> or <code>@vocab</code> you can define that thing.<br/>
… our mission is to use json in the wild where this is a pattern of usage<br/>
Benjamin Young: You said exactly what I was typing.<br/>
… it would be good to use an example from jsonapi<br/>
Gregg Kellogg: jsonapi examples are quite long, with a lot of nesting<br/>
… we have a test case from jsonapi<br/>
… may be too long for here. But may be good for best practices document.<br/>
… It would overly complicated the spec to include here.<br/>
Benjamin Young: This solves the jsonapi case by aliasing included to <code>@included.</code><br/>
Gregg Kellogg: Yes, you can have multiple properties that have multiple aliases.<br/>
… included can be a nested object<br/>
Ivan Herman: Can we talk about things that go to the primer?<br/>
Gregg Kellogg: What to do with example?<br/>
Ivan Herman: Switch the order of problem of Rob. In the primer we will have to spend more words on the fact that there are different things that can be used to do the same thing.<br/>
… We must have a primer.<br/>
… The current doc is already huge.<br/>
Benjamin Young: We need distinction in the main spec explaining diff between included and graph<br/>
Ivan Herman: To be honest, at this level there is no real diff between the examples.<br/>
… this is a side-effect with included.<br/>
… we should not fiddle around with that<br/>
Benjamin Young: The graph foundation exists in flattened output, and this won’t go away.<br/>
… this needs clarification<br/>
Gregg Kellogg: The use of included on its own is a by-product of the feature.<br/>
… it does not need its own description in the spec<br/>
… There are use cases where that can be useful<br/>
… That better lies in a non-normative text.<br/>
Benjamin Young: The bush usage would go to the primer<br/>
… focus of the text would go back to inter-document referencing.<br/>
Ivan Herman: We are falling back to other extreme that I don’t agree with<br/>
… we are hiding a feature of included<br/>
… In the inclusion part we should mention the alternative representation of bushes.<br/>
… because current graph-based bushes are a hack<br/>
… It has been around for a while, but we still should mention it.<br/>
Benjamin Young: This is an accidental feature since recently.<br/>
… it is essential to flattened output. I don’t see it as a hack.<br/>
… For non-turtle/trig users.<br/>
Ivan Herman: For semweb folks you don’t care it is a hack.<br/>
… we get into taste issues<br/>
… I don’t want to hide it.<br/>
Proposed resolution: focus <code>@included</code> text and example on original inclusion use case; mention value of it as an <code>@graph</code> replacement for bushes–and reference primer for further reading (Benjamin Young)<br/>
Ivan Herman: +0.5<br/>
Gregg Kellogg: +1<br/>
Benjamin Young: +1<br/>
Tim Cole: +1<br/>
Ruben Taelman: +1<br/>
Ivan Herman: If we have a primer, a reference can be put into it in CR<br/>
Resolution #2: focus <code>@included</code> text and example on original inclusion use case; mention value of it as an <code>@graph</code> replacement for bushes–and reference primer for further reading<br/>
Ivan Herman: include or included?<br/>
Benjamin Young: People may expect a URI to be included. But this is wrong.<br/>
Ruben Taelman: what was the keyword in JSON API?<br/>
Gregg Kellogg: included<br/>
Ruben Taelman: in that case, it might be helpful to be consistent with that<br/>
Tim Cole: +1 for <code>@included</code><br/>
Proposed resolution: close issue #19 with merger of <code>@included</code> related PRs (Benjamin Young)<br/>
Tim Cole: +1<br/>
Ivan Herman: +1<br/>
Ruben Taelman: +1<br/>
Benjamin Young: +1<br/>
Gregg Kellogg: +1<br/>
Resolution #3: close issue #19 with merger of <code>@included</code> related PRs<br/>
4<br/>
</details>

**Pull request merged**